File: server/romanize.py
#!/usr/bin/env python3
"""Romanize subtitle lines so a learner can see how the source is pronounced.

- Chinese  -> pinyin with tone marks, one syllable per character (pypinyin)
- Japanese -> Hepburn romaji, per word chunk (pykakasi)
- other non-Latin scripts (Korean, Cyrillic, Greek, Arabic, Hebrew, Devanagari,
  Thai, …) -> best-effort transliteration, per word (unidecode)
- Latin-script languages -> nothing (a romanization would just strip accents)

Reads one JSON object on stdin: {"lang": str, "lines": [str, ...]}.
Writes one JSON object on stdout: {"tokens": [line, ...]} where each `line` is a
list of [base, pron] pairs. Concatenating every `base` reconstructs the line, so
the frontend can stack `pron` directly over the character(s) it belongs to
(ruby/furigana style). `pron` is "" for anything not romanized (punctuation,
spaces, already-Latin text).
"""

import json
import re
import sys

# Languages already written in the Latin script: a pronunciation guide adds
# nothing. (Mirrors the Latin-script codes in server/translate.py.)
LATIN = {
    "af", "az", "ca", "cs", "da", "nl", "en", "eo", "et", "fi", "fr", "de",
    "hu", "id", "ga", "it", "lv", "lt", "ms", "nb", "pl", "pt", "ro", "sk",
    "sl", "es", "sv", "tl", "tr", "vi",
}

_HAN = r"㐀-䶿一-鿿豈-﫿"
_HAN_RUN = re.compile(f"[{_HAN}]+|[^{_HAN}]+")


def pinyin_tokens(text):
    from pypinyin import Style, pinyin

    tokens = []
    for run in _HAN_RUN.findall(text):
        if re.match(f"[{_HAN}]", run):
            # Pass the whole Han run so pypinyin can disambiguate polyphones from
            # context, then pair each syllable back with its character.
            sylls = pinyin(run, style=Style.TONE, errors="default")
            for ch, syl in zip(run, sylls):
                tokens.append([ch, syl[0]])
        else:
            tokens.append([run, ""])
    return tokens


def romaji_tokens(text):
    import pykakasi

    kks = pykakasi.kakasi()
    tokens = []
    for seg in kks.convert(text):
        orig, hepburn = seg["orig"], seg["hepburn"]
        # Blank the reading for ASCII/punctuation passthrough so we don't stack
        # text over itself.
        tokens.append([orig, "" if orig == hepburn else hepburn])
    return tokens


def translit_tokens(text):
    from unidecode import unidecode

    tokens = []
    for run in re.findall(r"\w+|\W+", text, re.UNICODE):
        roman = unidecode(run).strip() if re.search(r"\w", run, re.UNICODE) else ""
        tokens.append([run, "" if roman == run else roman])
    return tokens


def get_tokenizer(lang):
    lang = (lang or "").lower()
    if lang in ("zh", "zh-cn", "zh-tw", "chinese"):
        return pinyin_tokens
    if lang in ("ja", "japanese"):
        return romaji_tokens
    if lang in LATIN:
        return None
    return translit_tokens


def main():
    req = json.load(sys.stdin)
    lines = req.get("lines") or []
    tokenize = get_tokenizer(req.get("lang"))

    if tokenize is None:
        out = [[] for _ in lines]
    else:
        out = []
        for line in lines:
            try:
                out.append(tokenize(str(line)))
            except Exception:
                out.append([[str(line), ""]])

    json.dump({"tokens": out}, sys.stdout, ensure_ascii=False)
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: server/romanize.mjs
import path from "node:path";
import { fileURLToPath } from "node:url";
import { readJsonBody, runCommand, sendJson } from "./util.mjs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PYTHON_BIN = path.join(__dirname, "..", ".venv", "bin", "python");
const ROMANIZE_SCRIPT = path.join(__dirname, "romanize.py");

// POST /api/romanize  { language, lines: [string] } -> { tokens: [line] }
// Each `line` is a list of [base, pron] pairs so the client can stack the
// pronunciation over the character(s) it belongs to (ruby style). Best-effort:
// any failure yields an empty list so the page never breaks.
export async function handleRomanize(req, res) {
  const body = await readJsonBody(req);
  const lang = String(body.language || body.lang || "").trim();
  const lines = Array.isArray(body.lines) ? body.lines.map((l) => String(l)) : [];
  if (!lines.length) {
    sendJson(res, 200, { tokens: [] });
    return;
  }
  try {
    const result = await runCommand(PYTHON_BIN, [ROMANIZE_SCRIPT], {
      timeoutMs: 60_000,
      input: JSON.stringify({ lang, lines }),
    });
    const line = result.stdout.trim().split("\n").filter(Boolean).at(-1) || "{}";
    const data = JSON.parse(line);
    sendJson(res, 200, { tokens: data.tokens || [] });
  } catch (err) {
    console.error("Romanize failed:", err.message);
    sendJson(res, 200, { tokens: [] });
  }
}
